contracts/interactiveAdapters/ousd/OusdInteractiveAdapter.sol

```diff
@@ -0,0 +1,101 @@
+pragma experimental ABIEncoderV2;
+
+import { ERC20 } from "../../interfaces/ERC20.sol";
+import { SafeERC20 } from "../../shared/SafeERC20.sol";
+import { TokenAmount } from "../../shared/Structs.sol";
+import { ERC20ProtocolAdapter } from "../../adapters/ERC20ProtocolAdapter.sol";
+import { InteractiveAdapter } from "../InteractiveAdapter.sol";
+import { OusdVault } from "../../interfaces/OusdVault.sol";
+import { OusdToken } from "../../interfaces/OusdToken.sol";
+
+/**
+ * @title Interactive adapter for Balancer (liquidity).
+ * @dev Implementation of InteractiveAdapter abstract contract.
+ * @author Domen Grabec <domen@originprotocol.com>
+ */
+contract OusdInteractiveAdapter is InteractiveAdapter, ERC20ProtocolAdapter {
+    using SafeERC20 for ERC20;
+    address internal constant OUSD = 0x2A8e1E676Ec238d8A992307B495b45B3fEAa5e86;
+
+    /**
+     * @notice Deposits tokens to the Balancer pool.
+     * @param tokenAmounts Array with one element - TokenAmount struct with
+     * token address, token amount to be deposited, and amount type.
+     * @param data ABI-encoded additional parameter
+     * @return tokensToBeWithdrawn Array with one element - ousd address.
+     * @dev Implementation of InteractiveAdapter function.
+     */
+    function deposit(TokenAmount[] calldata tokenAmounts, bytes calldata data)
+        external
+        payable
+        override
+        returns (address[] memory tokensToBeWithdrawn)
+    {
+        require(tokenAmounts.length == 1, "OIA: should be 1 tokenAmount[1]");
+        address vaultAddress = OusdToken(OUSD).vaultAddress();
+
+        tokensToBeWithdrawn = new address[](1);
+        tokensToBeWithdrawn[0] = OUSD;
+
+        address token = tokenAmounts[0].token;
+        uint256 amount = getAbsoluteAmountDeposit(tokenAmounts[0]);
+        ERC20(token).safeApproveMax(vaultAddress, amount, "OIA");
+
+        // solhint-disable-next-line no-empty-blocks
+        try OusdVault(vaultAddress).mint(token, amount, 0) {} catch Error(string memory reason) {
+            revert(reason);
+        } catch {
+            revert("OIA: deposit fail");
+        }
+    }
+
+    /**
+     * @notice Withdraws tokens from the Balancer pool.
+     * @param tokenAmounts Array with one element - TokenAmount struct with
+     * OUSD token address, OUSD token amount to be redeemed, and amount type.
+     * @param data ABI-encoded additional parameter
+     * @return tokensToBeWithdrawn Array with one element - destination token address.
+     * @dev Implementation of InteractiveAdapter function.
+     */
+    function withdraw(TokenAmount[] calldata tokenAmounts, bytes calldata data)
+        external
+        payable
+        override
+        returns (address[] memory tokensToBeWithdrawn)
+    {
+        require(tokenAmounts.length == 1, "OIA: should be 1 tokenAmount[2]");
+
+        address toToken = abi.decode(data, (address));
+
+        // TODO: unknown which tokens will be received upon redemption
```

Comment:
Yeah that would be convenient. Unfortunately vault's redeem function doesn't return any value: 
https://github.com/OriginProtocol/origin-dollar/blob/master/contracts/contracts/vault/VaultCore.sol#L158

Reply:
Could that be a problem? I mean does Zerion absolutely need the information about token redemption?

Reply:
it's enough to add
```
address vaultAddress = OusdToken(token).vaultAddress();
tokensToBeWithdrawn = OusdVault(vaultAddress).getAllAssets();
```

Reply:
Aha I see thanks, altered as proposed.

To avoid any future confusion, when redeeming Vault won't necessarily return all 3 stablecoins. It might, but can also return only 1 or 2.

Reply:
is there a way to understand how many different tokens will be returned, 1 2 or 3 and which exactly? Returning tokens addresses which are not actually returns increases gas very slightly, it will check whether are tokens are actually returned, and if no -- do nothing))

Reply:
The closest thing would be the [calculateRedeemOutputs](https://github.com/OriginProtocol/origin-dollar/blob/master/contracts/contracts/vault/VaultCore.sol#L509) that is used by the Dapp to show the estimated stablecoin amounts returned.

If no `mint` / `redeem` calls happen between the  `calculateRedeemOutputs` function call and the `redeem` call those amounts will be exact. If someone else calls mint/redeem between these 2 function calls those estimated will be off.

Reply:
I think we can leave it as is

Reply:
👍